### PR TITLE
Support hierarchical paths for repos

### PR DESCRIPTION
**Pull request by g4s8**
#178 - support two different repository layout formats:
 - `flat` - no repository hierarchy
 - `org` - two level hierarchy: first level is organization name, second part is repository name

See `RepoLayout` javadoc for more details.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#187](https://github.com/artipie/artipie/issues/187) is already in scope

**Comment by 0crat** [comment]
This pull request [#187](https://github.com/artipie/artipie/issues/187) is assigned to @brastak[/z](https://www.0crat.com/u/brastak), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/cef1459c-ff29-492a-8744-d9bd08e635cb); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; @brastak[/z](https://www.0crat.com/u/brastak) is not a member of this project yet, but they can request to join, as [§1](http://www.zerocracy.com/policy.html#1) explains; there will be no monetary reward for this job

File: src/main/java/com/artipie/repo/OrgLayout.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import com.artipie.http.slice.SliceSimple;
+import com.jcabi.log.Logger;
+import io.vertx.reactivex.core.Vertx;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Pattern;
+import org.cactoos.scalar.Unchecked;
+
+/**
+ * Flat repositories layout.
```
Comment (Sammers21) [inline src/main/java/com/artipie/repo/OrgLayout.java:44]: The Javadoc has been copied from `FlatLayout`

File: src/main/java/com/artipie/Settings.java [hunk 2]
```diff
@@ -80,13 +92,18 @@ public Fake(final Storage storage) {
         }
 
         @Override
-        public Storage storage() throws IOException {
```
Comment (Sammers21) [inline src/main/java/com/artipie/Settings.java:83]: How those changes are related to the Pull Request?

**Comment by Sammers21** [comment]
Tests and documentation on how to use the `Org` layout are missing.

File: src/main/java/com/artipie/repo/RepoLayout.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Repositories layout.
+ * <p>
+ * This object resolves repository HTTP endpoints for specified URL.
+ * Artipie supports two primary layout configurations:
+ * <ul>
+ *     <li>Flat layout - Artipie doesn't use any structural layout, all repositories
+ *     are located at the root of base path, e.g. URI {@code https://central.artipie.com/maven}
+ *     accesses {@code maven} repository.</li>
+ *     <li>User based - Artipie installation has multiple namespaces/organizations/users,
```
Comment (Sammers21) [inline src/main/java/com/artipie/repo/RepoLayout.java:37]: `Org layout` and `User based` are different terms. I'd suggest sticking with only one, so the Javadoc would be consistent with the actual code

File: src/main/java/com/artipie/Settings.java [hunk 4]
```diff
@@ -35,6 +38,7 @@
  * @since 0.1
  */
 public interface Settings {
+
```
Comment (Sammers21) [inline src/main/java/com/artipie/Settings.java:41]: Is this change somehow related to the Pull Request?

File: src/main/java/com/artipie/repo/OrgLayout.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+ * @see RepoLayout
+ * @since 0.4
+ * @checkstyle ReturnCountCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+public final class OrgLayout implements RepoLayout {
```
Comment (Sammers21) [inline src/main/java/com/artipie/repo/OrgLayout.java:50]: Except for one field `REPO_PREF`, how `OrgLayout` is different from `FlatLayout`? I could not find any differences.

File: src/main/java/com/artipie/repo/OrgLayout.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Repository path prefix.
+     */
+    private static final Pattern REPO_PREF =
+        Pattern.compile("/(?:[a-zA-Z0-9]+)/(?:[a-zA-Z0-9]+)(/.*)");
```
Comment (brastak) [inline src/main/java/com/artipie/repo/OrgLayout.java:56]: @g4s8 I think, this pattern is too restricted: how about dashes ('-'), underscores ('_') and some other symbols? Valid path symbols can be looked [here](https://stackoverflow.com/questions/4669692/valid-characters-for-directory-part-of-a-url-for-short-links)

File: src/main/java/com/artipie/YamlSettings.java [hunk 7]
```diff
@@ -112,6 +116,23 @@ public Storage storage() throws IOException {
         return res;
     }
 
+    @Override
+    public RepoLayout layout(final Vertx vertx) throws IOException {
+        final String layout = Yaml.createYamlInput(this.content)
+            .readYamlMapping()
+            .yamlMapping(YamlSettings.META)
+            .string("layout");
```
Comment (brastak) [inline src/main/java/com/artipie/YamlSettings.java:124]: @g4s8 We need test for new param and documentation (please, add `todo`s)

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 5 above)
Comment (Sammers21) [reply src/main/java/com/artipie/repo/OrgLayout.java:50]: I think it is a code duplication which should be fixed.

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 6 above)
Comment (Sammers21) [inline src/main/java/com/artipie/repo/OrgLayout.java:56]: The only difference between this pattern and the one at `FlatLayout.REPO_PREF` is the minimum amount of `/` symbols and parts in between them, 2 in case of `FlatLayout` and 3 in this case. This also could be decomposed. An example:
`Pattern REPO_PREF = new PathLevel(3)`

**Comment by Sammers21** [comment]
My overall impression is that a small change requires a huge amount of code changes, which is definitely a bad signal

File: src/main/java/com/artipie/Settings.java (hunk 2 above)
Comment (g4s8) [reply src/main/java/com/artipie/Settings.java:83]: @Sammers21 it's just a minor cosmetic change in related class

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 5 above)
Comment (g4s8) [reply src/main/java/com/artipie/repo/OrgLayout.java:50]: @Sammers21 `OrgLayout` reads two first part of URI to find repository config and to specify storage prefix, `FlatLayout` uses only first. But it can be refactored later.

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 6 above)
Comment (g4s8) [reply src/main/java/com/artipie/repo/OrgLayout.java:56]: @brastak we have some restrictions for repository names, see https://github.com/artipie/artipie
>  Make sure that the name of your config file is the name of repository you are going to host, and its name matches `[a-z0-9_]{3,32}`

So, I've changed it to `[a-zA-Z0-9_]`

**Comment by Sammers21** [comment]
@g4s8, What about higher levels of hierarchy or let it be dynamically defined by a user?
If a hierarch of repository configs is flat:
```
├── gem.yaml
└── maven.yaml
```
let the layout be flat as well.

But what if config storage structure is the following:
```
├── depatmet-1
│   └── dev
│       ├── gem.yaml
│       └── maven.yaml
└── depatmet-2
    └── dev
        ├── gem.yaml
        └── maven.yaml
```

Should the pathes for the repositories be `depatmet-1/dev/gem`,  `depatmet-1/dev/maven`, etc...? I guess the answer is Yes.

I think that paths should be dependent on storage structure and dynamically defined by a user.

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 6 above)
Comment (g4s8) [reply src/main/java/com/artipie/repo/OrgLayout.java:56]: @Sammers21 `Pattern` is a `final` class and can't be extended. Also, we may consider changing the format of first or second part here, so it'll require more changes than using pattern string. I'd keep it as is while the format is not established as a requirement.

Comment (Sammers21) [reply src/main/java/com/artipie/repo/OrgLayout.java]: @g4s8, how about this?:
`Pattern REPO_PREF = Pattern.compile(new PathLevel(3))`

File: src/main/java/com/artipie/Settings.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/Settings.java:83]: @g4s8, move this change to a separated Pull Request with `minor changes` since this change has no intersections with this Pull Request goals.

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 5 above)
Comment (Sammers21) [reply src/main/java/com/artipie/repo/OrgLayout.java:56]: @g4s8, where is the puzzle for that?

File: src/main/java/com/artipie/repo/FlatLayout.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+ * @since 0.4
+ * @checkstyle ReturnCountCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+public final class FlatLayout implements RepoLayout {
+
+    /**
+     * Repository path prefix.
+     */
+    private static final Pattern REPO_PREF = Pattern.compile("/(?:[a-zA-Z0-9_]+)(/.*)");
```
Comment (Sammers21) [inline src/main/java/com/artipie/repo/FlatLayout.java:60]: `(?:[a-zA-Z0-9_]+)` repeated 3 times, decompose it somehow so you would not do the same changes 3 times like you did it in 448782ad2de072decf0cade1550a2ca91b051fd5

Comment (g4s8) [reply src/main/java/com/artipie/repo/FlatLayout.java:60]: @Sammers21 see https://github.com/artipie/artipie/pull/187#discussion_r433934152 we can do it later when the format of these parts will be stable

File: src/main/java/com/artipie/repo/OrgLayout.java (hunk 6 above)
Comment (brastak) [reply src/main/java/com/artipie/repo/OrgLayout.java]: @g4s8 We need to validate then when instantiate adapters from config or to make restriction weaker. Examples in our `README.md` violate this constraint (`npm-proxy`, `my-php`)

Comment (g4s8) [reply src/main/java/com/artipie/repo/OrgLayout.java]: @brastak done, fixed

**Comment by 0pdd** [comment]
@g4s8 the puzzle [#191](https://github.com/artipie/artipie/issues/191) is still not solved.

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @brastak[/z](https://www.0crat.com/u/brastak), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Job was finished in 20 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by sereshqua** [comment]
@Sammers21 please make sure you start all your comments with the name of the user they are referred to, [see](https://github.com/artipie/artipie/pull/187#issuecomment-637580460)

**Comment by Sammers21** [comment]
@sereshqua, please pay attention that my comment was ignored and merge proceeded with reviewers approve.

**Comment by sereshqua** [comment]
@Sammers21  - [see](https://github.com/artipie/artipie/pull/187#discussion_r433948855), you received a response

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Comment by Sammers21** [comment]
@sereshqua - [see](https://github.com/artipie/artipie/pull/187#issuecomment-637589672), i have not

**Comment by 0pdd** [comment]
@g4s8 the only puzzle [#191](https://github.com/artipie/artipie/issues/191) is solved here.
